Signal Red Hat keys presence in the cluster.status

petrkotas opened the pull request:
### Which issue this PR addresses:

It is first half as per request, signalling functionality and mutating functionality is split to simplify review process.

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/8749451/

### What this PR does / why we need it:
Add Red Hat keys list to cluster.status

Add a checks that parses pull-secret from openshift-config
namespace and extracts the list of present Red Hat keys.
Keys are presented in a cluster.status.RedHatKeysPresent field.

Signed-off-by: Petr Kotas <pkotas@redhat.com>

---

Incorporates changes requested in: #1347 


### Test plan for issue:
Unit tests added

### Is there any documentation that needs to be updated for this PR?
Documented inline.

mjudeikis commented:
@petrkotas is the case that if we dont have any case we will have condition with no values? Do you know what is the view in this case?

petrkotas commented:
This is a good question that made me think whether I really understand how the status works. So I:

* checked that I am omitting an empty field: https://github.com/Azure/ARO-RP/pull/1425/files#diff-fcc49fbb4570ce111fb58894129934b7dec62a17ee96c951ef77a3228e61344fR66 
* updated the tests to ensure field is really nil: https://github.com/Azure/ARO-RP/pull/1425/files#diff-a6c1b61302c569ca971865562fd5788b0727c6d1b60b74ad3bbda826df983f1aR77
* updated tests to double check that field is not present: https://github.com/Azure/ARO-RP/pull/1425/files#diff-a6c1b61302c569ca971865562fd5788b0727c6d1b60b74ad3bbda826df983f1aR264
* triple checked manually:

With keys:
![Screenshot from 2021-04-13 13-32-39](https://user-images.githubusercontent.com/8687113/114546659-b6efea00-9c5d-11eb-93ef-b0ed6376e67f.png)

Without keys:
![Screenshot from 2021-04-13 13-33-12](https://user-images.githubusercontent.com/8687113/114546704-c7a06000-9c5d-11eb-86ff-70b4346de630.png)

So I am fairly confident there will be no empty list shown in the status.

petrkotas commented:
Rebased

mjudeikis commented on pkg/operator/controllers/pullsecret/pullsecret_controller_test.go:
```diff
@@ -145,30 +234,412 @@ func TestPullSecretReconciler(t *testing.T) {
 
-			if string(s.Data[corev1.DockerConfigJsonKey]) != tt.want {
-				t.Error(string(s.Data[corev1.DockerConfigJsonKey]))
+			status := *&arov1alpha1.ClusterStatus{}
```
This does not look right...

mjudeikis commented on pkg/operator/controllers/pullsecret/pullsecret_controller_test.go:
```diff
@@ -145,30 +234,412 @@ func TestPullSecretReconciler(t *testing.T) {
-			if string(s.Data[corev1.DockerConfigJsonKey]) != tt.want {
-				t.Error(string(s.Data[corev1.DockerConfigJsonKey]))
+			status := *&arov1alpha1.ClusterStatus{}
+			err = json.Unmarshal(statusBytes, &status)
```
`err` never used?

mjudeikis commented on pkg/operator/controllers/pullsecret/pullsecret_controller_test.go:
```diff
@@ -145,30 +234,412 @@ func TestPullSecretReconciler(t *testing.T) {
 
-			if s.Type != corev1.SecretTypeDockerConfigJson {
-				t.Error(s.Type)
+			statusBytes := []byte{}
+			statusBytes, err = json.Marshal(&cluster.Status)
```
`statusBytes, err := json.Marshal(&cluster.Status)`

mjudeikis commented on pkg/operator/controllers/pullsecret/pullsecret_controller_test.go:
```diff
@@ -109,7 +139,56 @@ func TestPullSecretReconciler(t *testing.T) {
+			}, &corev1.Secret{Data: map[string][]byte{
+				corev1.DockerConfigJsonKey: []byte(`{"auths":{"arosvc.azurecr.io":{"auth":"ZnJlZDplbnRlcg=="},"registry.redhat.io":{"auth":"ZnJlZDplbnRlcg=="}}}`),
+			}}),
+			arocli: newFakeAro(
```
These are ok, but this is not the pattern we use around the codebase. We usually do this via pointer functions: `func(f *testdatabase.Fixture)` where we define function to modify the input based on test requirements. So you end-up passing same static model into function to adopt it to the test case requirements.

This way we keep the same structure and is easier to read. Now here some usecases has `arocli:   baseCluster,` and some `arocli: newFakeAro(stuff)` this is doing same as pointer functions but without consistency. 


Example:
```
fixture: func(f *testdatabase.Fixture) {
				f.AddOpenShiftClusterDocuments(&api.OpenShiftClusterDocument{
					Key: strings.ToLower(resourceID),
					OpenShiftCluster: &api.OpenShiftCluster{
						ID:       resourceID,
						Name:     "resourceName",
						Type:     "Microsoft.RedHatOpenShift/OpenShiftClusters",
						Location: "location",
						Properties: api.OpenShiftClusterProperties{
							ProvisioningState: api.ProvisioningStateCreating,
						},
					},
				})
				f.AddSubscriptionDocuments(&api.SubscriptionDocument{
					ID: mockSubID,
				})
			},
			checker: func(c *testdatabase.Checker) {
				c.AddOpenShiftClusterDocuments(&api.OpenShiftClusterDocument{
					Key: strings.ToLower(resourceID),
					OpenShiftCluster: &api.OpenShiftCluster{
						ID:       resourceID,
						Name:     "resourceName",
						Type:     "Microsoft.RedHatOpenShift/OpenShiftClusters",
						Location: "location",
						Properties: api.OpenShiftClusterProperties{
							ProvisioningState: api.ProvisioningStateCreating,
							Install: &api.Install{
								Phase: api.InstallPhaseBootstrap,
							},
						},
					},
				})
			},
```

mjudeikis reviewed (changes requested):
Few changes :)

Makdaam commented on pkg/operator/controllers/pullsecret/pullsecret_controller.go:
```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	create := userSecret == nil
+	remove := false
+
+	// userSecret can happen to have broken type, which means it have to be recreated
```
nit "has to be recreated"

Makdaam commented on pkg/operator/controllers/pullsecret/pullsecret_controller.go:
```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+
+	if remove {
+		// unfortunately the type field is immutable.
```
should this comment be somewhere near line 150?
Or is it the only scenario when the secret will be removed ever?

petrkotas replied:
you are right, this make sense to mention on the line where it is set. The recreation is required when the `.type` field is broken. This field is immutable. I have moved the comment.

Makdaam commented on pkg/operator/controllers/pullsecret/pullsecret_controller.go:
```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		return nil, err
+	}
+
+	if !create && !update {
+		return userSecret, nil
+	}
```
I'd move this as high as possible, I'm guessing line 155. with a comment above that the "create" and "remove" control flow booleans don't change below this point.

petrkotas replied:
After a discussion with @Makdaam I have decided to change the logic flow to make it more maintainable. Now it does not handle the complexity of decision whether to recreate or just create.

Makdaam reviewed (changes requested):
Requesting some minor changes in fixAndUpdateGlobalPullSecret, I'm still thinking if it can be simplified more (with less if create) but I don't think so.

nilsanderselde commented on pkg/operator/controllers/pullsecret/pullsecret_controller.go:
```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		return nil, errors.New("nil operator secret, cannot verify userData integrity")
+	}
+
+	create := userSecret == nil
+	remove := false
+
+	// userSecret can happen to have broken type, which means it have to be recreated
+	// with proper type
+	if userSecret != nil && (userSecret.Type != corev1.SecretTypeDockerConfigJson || userSecret.Data == nil) {
+		create = true
+		remove = true
+	}
```
I got a little confused here. I didn't understand at first how both `create` and `remove` could be true.

`remove` is true if and only if "`create` is true because the secret existed but it was bad."

Maybe a better name for `remove` would be `replace`.

Additionally, we could make the logic easier to read like so:

```suggestion
	var create, replace bool
	if userSecret == nil {
		create = true
	} else if userSecret.Type != corev1.SecretTypeDockerConfigJson || userSecret.Data == nil {
		create, replace = true, true
	}
```

petrkotas replied:
I have actually changed the whole function to make it more maintainable. The thought is to make the flow easier to read, but sacrifice one deletion call to the API. 
The delete happens only when there is a change, so there is no significant performance penalty.

nilsanderselde commented on pkg/operator/controllers/pullsecret/pullsecret_controller.go:
```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		return foundKeys, err
+	}
+
+	if parsedKeys == nil {
+		return foundKeys, nil
+	}
+
+	for _, rhKey := range rhKeys {
+		if v := parsedKeys[rhKey]; len(v) > 0 {
+			foundKeys = append(foundKeys, rhKey)
+		}
+	}
+
+	return foundKeys, nil
```
We're returning `foundKeys, nil` here regardless, so this could be simplified:

```suggestion
	if parsedKeys != nil {
		for _, rhKey := range rhKeys {
			if v := parsedKeys[rhKey]; len(v) > 0 {
				foundKeys = append(foundKeys, rhKey)
			}
		}
	}
	return foundKeys, nil
```

petrkotas replied:
ah good eye, thanks

nilsanderselde reviewed (changes requested):
Looks good - just have a few suggestions

petrkotas commented:
I have updated the code and did the rebase.

nilsanderselde reviewed (commented):
LGTM. `ensureGlobalPullSecret` is much easier to follow now.
